clients/apple/Shared/Stateful Logic/BillingService.swift

```diff
@@ -0,0 +1,197 @@
+    var purchaseResult: PurchaseResult? = nil
+    @Published var cancelSubscriptionResult: CancelSubscriptionResult? = nil
+    
+    var makingPurchaseAttempt = false
+    
+    
+    init(_ core: LockbookApi) {
+        self.core = core
+        updateListenerTask = listenForTransactions()
+
+        Task {
+            await requestProducts()
+        }
+    }
+    
+    deinit {
+        updateListenerTask?.cancel()
+    }
+    
+    func listenForTransactions() -> Task<Void, Error> {
+        return Task.detached {
+            for await verification in Transaction.updates {
+                do {
+                    let transaction = try self.checkVerified(verification)
+                    switch self.core.getUsage() {
+                    case .success(let usages):
+                        if usages.dataCap.exact == 1000000 {
+                            if let receipt = self.getReceipt(), !self.makingPurchaseAttempt, transaction.id == transaction.originalID {
+                                guard let appAccountToken = transaction.appAccountToken else {
+                                    throw StoreError.failedVerification
+                                }
+                                
+                                let result = self.core.upgradeAccountAppStore(originalTransactionId: String(transaction.originalID), appAccountToken: appAccountToken.uuidString.lowercased(), encodedReceipt: receipt)
+                                
+                                switch result {
+                                case .success(_):
+                                    await transaction.finish()
+                                    self.purchaseResult = .success
+                                case .failure(let error):
+                                    DI.errors.handleError(error)
+                                }
+                            }
+                        } else {
+                            await transaction.finish()
+                        }
+                    case .failure(let error):
+                        DI.errors.handleError(error)
+                    }
+                } catch {
+                    print("Transaction failed verification")
+                }
+            }
+        }
+    }
+
+    
+    func purchasePremium() async throws -> PurchaseResult? {
+        let accountToken = UUID()
+        let purchaseOpt: Set<Product.PurchaseOption> = [Product.PurchaseOption.appAccountToken(accountToken)]
+        
+        do {
+            guard let monthlySubscription = maybeMonthlySubscription else {
+                throw StoreError.noProduct
+            }
+            let result = try await monthlySubscription.purchase(options: purchaseOpt)
+            SKReceiptRefreshRequest().start()
+            
+            switch result {
+            case .success(let verification):
+                while true {
+                    if let receipt = getReceipt() {
```

Comment:
Can there be no receipt? Wouldn't this inf loop if no receipt?

Reply:
The receipt should [be generated after a purchase](https://developer.apple.com/documentation/appstorereceipts/validating_receipts_with_the_app_store). It's just a matter of waiting.

Reply:
```
    func getReceipt() -> String? {
        if let appStoreReceiptURL = Bundle.main.appStoreReceiptURL,
            FileManager.default.fileExists(atPath: appStoreReceiptURL.path) {

            do {
                return try Data(contentsOf: appStoreReceiptURL, options: .alwaysMapped).base64EncodedString()
            }
            catch {
                return nil
            }
        }
        
        return nil
    }
    ```
    
I think waiting for a file to exist in a loop is fine, you should add a bit of a sleep regardless if you're going to do this. Can you link to a sample that does this? Feels strange regardless. But you should have a timeout, if some assumption of yours is wrong this will infinite loop.

1. handle the errors in the catch better
2. sleep and timeout